hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/CleanerChore.java

```diff
@@ -322,140 +330,154 @@ public boolean getEnabled() {
-      } catch (IOException ioe) {
-        LOG.warn("failed to get FileStatus for contents of '" + dir + "'", ioe);
-        return false;
+        // Mark the result set process finished and notify the waiting get method.
+        this.resultSet = true;
+        this.notify();
       }
+    }
 
-      boolean allFilesDeleted = true;
-      if (!files.isEmpty()) {
-        allFilesDeleted = deleteAction(new Action<Boolean>() {
-          @Override
-          public Boolean act() throws IOException {
-            return checkAndDeleteFiles(files);
-          }
-        }, "files");
+    synchronized boolean get() throws Exception {
+      while (!resultSet) {
+        wait();
+      }
+      return result;
+    }
+  }
+
+  /**
+   * Attempts to clean up a directory(its subdirectories, and files) in a
+   * {@link java.util.concurrent.ThreadPoolExecutor} concurrently. We can get the final result by
+   * calling result.get().
+   */
+  private void traverseAndDelete(final Path dir, final boolean root, final AsyncResult result) {
+    try {
+      final Action<Boolean> curDirDeletion = new Action<Boolean>() {
+        @Override
+        public Boolean act() throws IOException {
+          return fs.delete(dir, false);
+        }
+      };
+
+      // Step.1: List all files under the given directory.
+      List<FileStatus> allPaths = Arrays.asList(fs.listStatus(dir));
+      final List<FileStatus> subDirs = new ArrayList<>();
+      final List<FileStatus> files = new ArrayList<>();
+      for (FileStatus status : allPaths) {
+        if (status.isDirectory()) {
+          subDirs.add(status);
+        } else if (status.isFile()) {
+          files.add(status);
+        }
       }
 
-      boolean allSubdirsDeleted = true;
-      if (!subDirs.isEmpty()) {
-        final List<CleanerTask> tasks = Lists.newArrayListWithCapacity(subDirs.size());
-        for (FileStatus subdir : subDirs) {
-          CleanerTask task = new CleanerTask(subdir, false);
-          tasks.add(task);
-          task.fork();
+      // Step.2: Try to delete all the deletable files.
+      final boolean allFilesDeleted = files.isEmpty() || deleteAction(new Action<Boolean>() {
+        @Override
+        public Boolean act() throws IOException {
+          return checkAndDeleteFiles(files);
         }
-        allSubdirsDeleted = deleteAction(new Action<Boolean>() {
-          @Override
-          public Boolean act() throws IOException {
-            return getCleanResult(tasks);
-          }
-        }, "subdirs");
+      }, "files", dir);
+
+      // Step.3: Start to traverse and delete the sub-directories.
+      if (subDirs.isEmpty()) {
+        // If no sub-directories, then just try to delete the current dir and finish the result.
+        boolean deleted = true;
+        if (allFilesDeleted && !root) {
+          deleted = deleteAction(curDirDeletion, "dir", dir);
+        }
+        result.set(deleted);
```

Comment:
There's no more sub dir, but if there're files remaining, we still return true? A bit confusing, i think it should be false.

Reply:
The deleted flag means whether we have handled the current  traverse & deletion successfully, not means the current dir is deleted successfuly, say if traversed all the subdirs & files successfully but still remain some sub-dirs there, the deleted should also be true.  Keep the same meaning as before. Maybe I need a more clearer javadoc.
Thanks.

Reply:
> * @return true if it deleted successfully, false otherwise

Doc of method `deleteAction`

Reply:
A `return true` of deletion of a dir should means all files and sub-dirs are deleted successfully? 

Then there's probability that there're file remaining(allFilesDeleted is false) inside this `if` branch?

Reply:
If all sub-files & sub-dirs are deleted successfully, but  failed to delete the current dir, the flag will also be false ... So I'd prefer to define the flag as whether we traverse & delete it succesfully (desipte whether there're some sub-files or dirs under there)